src/main/groovy/edu/oregonstate/mist/api/Resource.groovy

```diff
@@ -83,4 +105,90 @@ abstract class Resource {
                 details: properties.get('internalServerError.details')
         ))
     }
+
+    void setEndpointUri(URI endpointUri) {
+        this.endpointUri = endpointUri
+    }
+
+    /**
+     * Returns string url to use in pagination links.
+     *
+     * @param params
+     * @return
+     */
+    protected String getPaginationUrl(def params) {
+        def uriAndPath = endpointUri.toString() + uriInfo.getPath()
+        def nonNullParams = params.clone()
+        // convert pageVariable to page[variable]
+        nonNullParams["page[number]"] = nonNullParams['pageNumber']
+        nonNullParams["page[size]"] = nonNullParams['pageSize']
+        nonNullParams.remove('pageSize')
+        nonNullParams.remove('pageNumber')
+
+        // remove empty GET parameters
+        nonNullParams = nonNullParams.findAll { it.value } .collect { k, v -> "$k=$v" }
+
+        uriAndPath + "?" + nonNullParams.join('&')
+    }
+
+    /**
+     * Returns the value for an array parameter in the GET string.
+     *
+     * The JSONAPI format reserves the page parameter for pagination. This API uses page[size] and
+     * page[number].
+     * This function allows us to get just value for a specific parameter in an array.
+     *
+     * @param key
+     * @param index
+     * @param queryParameters
+     * @return
+     */
+    protected static String getArrayParameter(String key, String index,
+                                              MultivaluedMap<String, String> queryParameters) {
+        for (Map.Entry<String, List<String>> entry : queryParameters.entrySet()) {
+            // not an array parameter
+            if (!entry.key.contains("[") && !entry.key.contains("]")) {
+                continue
+            }
+
+            int a = entry.key.indexOf('[')
+            int b = entry.key.indexOf(']')
+
+            if (entry.key.substring(0, a).equals(key)) {
+                if (entry.key.substring(a + 1, b).equals(index)) {
+                    return entry.value?.get(0)
+                }
+            }
+        }
```

Comment:
Why do you have to iterate over the entire map? Can't you just do queryParameters.getFirst(key+'['+value+']')? For that matter, can't callers just do uriInfo.getQueryParameters().getFirst('page[number'])?